remote-build-coordinator/src/main/java/org/jboss/pnc/remotecoordinator/builder/RemoteBuildCoordinator.java

```diff
@@ -255,151 +299,194 @@ public BuildSetTask buildSet(BuildConfigurationSet buildConfigurationSet, User u
      * @return The new build set task
      * @throws CoreException Thrown if there is a problem initializing the build
      */
+    @Transactional // TODO test in the integration tests
+    @Retry(retryOn = ScheduleConflictExceptionWithAttachment.class)
+    @Fallback(BuildSetFallback.class)
     @Override
     public BuildSetTask buildSet(
             BuildConfigurationSet buildConfigurationSet,
             Map<Integer, BuildConfigurationAudited> buildConfigurationAuditedsMap,
             User user,
-            BuildOptions buildOptions) throws CoreException {
+            BuildOptions buildOptions) throws CoreException, BuildConflictException {
+
+        try {
+            Collection<BuildTaskRef> unfinishedTasks = taskRepository.getUnfinishedTasks();
+            checkAllRunning(buildConfigurationAuditedsMap.values(), unfinishedTasks);
 
-        synchronized (buildMethodLock) {
-            BuildSetTask buildSetTask = buildTasksInitializer.createBuildSetTask(
+            Graph<RemoteBuildTask> buildGraph = buildTasksInitializer.createBuildGraph(
                     buildConfigurationSet,
                     buildConfigurationAuditedsMap,
                     user,
                     buildOptions,
-                    this::buildRecordIdSupplier,
-                    taskRepository.getUnfinishedTasks());
-            updateBuildSetTaskStatus(buildSetTask, BuildStatus.BUILDING);
-
-            validateBuildConfigurationSetTask(buildConfigurationSet, buildOptions, buildSetTask);
-
-            build(buildSetTask);
+                    unfinishedTasks);
+            ScheduleResult scheduleResult;
+            try {
+                Long buildConfigSetRecordId = Sequence.nextId();
+                scheduleResult = validateAndRunBuilds(user, buildOptions, buildGraph, buildConfigSetRecordId);
+            } catch (ScheduleConflictException e) {
+                scheduleResult = new ScheduleResult(
+                        buildGraph,
+                        BuildCoordinationStatus.SYSTEM_ERROR,
+                        new BuildStatusWithDescription(
+                                BuildStatus.SYSTEM_ERROR,
+                                "Failed to schedule remote builds. " + e.getMessage()),
+                        Collections.emptySet());
+                throw new ScheduleConflictExceptionWithAttachment(
+                        e,
+                        buildConfigurationSet,
+                        user,
+                        buildOptions,
+                        scheduleResult);
+            }
+            return storeAndNotifyBuildSet(buildConfigurationSet, user, buildOptions, scheduleResult);
+        } catch (BuildConflictException e) {
+            log.warn("Conflicting build.", e);
+            throw e;
+        } catch (ScheduleConflictExceptionWithAttachment e) {
+            throw e;
+        } catch (Throwable e) {
+            String errorMessage = "Unexpected error while trying to schedule build set.";
```

Comment:
Shouldn't we throw exception on unexpected error? Do we return 500 to user if we fail unexpectedly?

Reply:
No, if we store the record we don't return 500. It's one or the other.

Reply:
@matejonnet  You are absolutely correct. Thanks. Do we return 500 when Rex is unavailable/request times out?

Reply:
As discussed off-line, storing the NO_REBUILD and throwing on SYSTEM_ERROR and conflicting build (cycle, empty)